backend/ipip/ipip.go

```diff
@@ -0,0 +1,182 @@
+		BackendType: backendType,
+	}
+
+	l, err := be.sm.AcquireLease(ctx, attrs)
+	switch err {
+	case nil:
+		n.SubnetLease = l
+	case context.Canceled, context.DeadlineExceeded:
+		return nil, err
+	default:
+		return nil, fmt.Errorf("failed to acquire lease: %v", err)
+	}
+	dev, err := be.configureIPIPDevice(n.SubnetLease)
+	if err != nil {
+		return nil, err
+	}
+	dev.directRouting = cfg.DirectRouting
+	n.dev = dev
+
+	return n, nil
+}
+
+func (be *IPIPBackend) configureIPIPDevice(lease *subnet.Lease) (*tunnelDev, error) {
+	link, err := netlink.LinkByName(tunnelName)
+	if err != nil {
+		// When modprobe ipip module, a tunl0 ipip device is created automatically per network namespace by ipip kernel module.
+		// It is the namespace default IPIP device with attributes local=any and remote=any.
+		// When receiving IPIP protocol packets, kernel will forward them to tunl0 as a fallback device
+		// if it can't find an option whose local/remote attribute matches their src/dst ip address more precisely.
+		// See https://github.com/torvalds/linux/blob/v4.13/net/ipv4/ip_tunnel.c#L85-L95 .
+
+		// So we have two options of creating ipip device, either rename tunl0 to flannel.ipip or create an new ipip device
+		// and set local attribute of flannel.ipip to distinguish these two devices.
+		// Considering tunl0 might be used by users, so choose the later option.
+		link = &netlink.Iptun{LinkAttrs: netlink.LinkAttrs{Name: tunnelName}, Local: be.extIface.IfaceAddr}
+		if err := netlink.LinkAdd(link); err != nil {
+			return nil, fmt.Errorf("failed to create tunnel %v: %v", tunnelName, err)
+		}
+	}
+	// flannel will never make this happen. This situation can only be caused by a user, so get them to sort it out.
+	if link.Type() != "ipip" {
+		return nil, fmt.Errorf("%v isn't an ipip mode device", tunnelName)
+	}
+	ipip, ok := link.(*netlink.Iptun)
+	if !ok {
+		return nil, fmt.Errorf("%s isn't an iptun device %#v", tunnelName, link)
+	}
+
+	// Don't set remote attribute making flannel.ipip an one to many tunnel device.
+	if (ipip.Local != nil && !ipip.Local.Equal(be.extIface.IfaceAddr)) || (ipip.Remote != nil && ipip.Remote.String() != "0.0.0.0") {
+		return nil, fmt.Errorf("local %v or remote %v of tunnel %s is not expected, please fix it and try again", ipip.Local, ipip.Remote, tunnelName)
+	}
+
+	// Due to the extra 20 byte IP header that the tunnel will add to each packet,
+	// MTU size for both the workload and tunnel interfaces should be 20 bytes less than the selected iface (specified with the --iface option).
+	expectMTU := be.extIface.Iface.MTU - 20
+	if expectMTU <= 0 {
+		return nil, fmt.Errorf("MTU %d of iface %s is too small for ipip mode to work", be.extIface.Iface.MTU, be.extIface.Iface.Name)
+	}
+	oldMTU := link.Attrs().MTU
+	if oldMTU > expectMTU || oldMTU == 0 {
+		glog.Infof("current MTU of %s is %d, setting it to %d", tunnelName, oldMTU, expectMTU)
+		err := netlink.LinkSetMTU(link, expectMTU)
```

Comment:
Does this need to explicitly set the link to down first?

Reply:
I couldn't find any reference about  this. I think changing MTU on the fly is ok. 

From http://linux-ip.net/html/tools-ip-link.html 
> # B.3.5. Using ip link set to change the MTU
> The remaining options to the ip link command cannot be used while the interface is in an UP state.